bindings/bind_py_structure_manager.cc

```diff
@@ -616,6 +618,196 @@ namespace rascal {
+            }
+          }
+
+          auto property_name{managers.get_calculator_name(calculator, true)};
+
+          const auto & property_ =
+              *managers[0]->template get_property<Prop_t>(property_name);
+          // assume inner_size is consistent for all managers
+          int inner_size{property_.get_nb_comp() / ThreeD};
+
+          math::Matrix_t features{};
+          auto n_rows{ThreeD * managers.template get_number_of_elements<Prop_t>(
+                                   property_name)};
+          size_t n_cols{all_keys.size() * inner_size};
+          features.resize(n_rows, n_cols);
+          features.setZero();
+
+          int i_row{0};
+          for (auto & manager : managers) {
+            const auto & property =
+                *manager->template get_property<Prop_t>(property_name);
+            auto n_rows_manager = property.size() * ThreeD;
+            property.fill_dense_feature_matrix_gradient(
+                features.block(i_row, 0, n_rows_manager, n_cols), all_keys);
+            i_row += n_rows_manager;
+          }
+
+          return features;
+        },
+        R"(Get the dense feature matrix associated with the calculator and
+        the collection of structures (managers) using the list of keys
+        provided. Only applicable when Calculator uses BlockSparseProperty.)");
+    manager_collection.def(
+        "get_atoms_for_predictions",
+        [](ManagerCollection_t & managers) {
+          if (managers.size() == 0) {
+            throw std::runtime_error(
+                R"(There are no structure to get features from)");
+          }
+          auto n_atoms{0};
+          for (auto & manager : managers) {
+            n_atoms += manager->size();
+          }
+
+          Eigen::Matrix<int, Eigen::Dynamic, 3> ij_mat{};
+          ij_mat.resize(n_atoms, 3);
+          ij_mat.setZero();
+          int i_center{0}, i_frame{0};
+          for (auto & manager : managers) {
+            for (auto center : manager) {
+              ij_mat(i_center, 0) = i_frame;
+              ij_mat(i_center, 1) = i_center;
+              ij_mat(i_center, 2) = center.get_atom_type();
+              i_center++;
+            }
+            i_frame++;
+          }
+          return ij_mat;
+        },
+        R"(Get informations necessary to the computation of predictions. It has
+        as many rows as the number representations and they correspond to the
+        index of the structure, the central atom and its atomic species.
+        np.array with shape (n_atoms, 3))");
+
+    manager_collection.def(
+        "get_neighbors_for_gradient",
+        [](ManagerCollection_t & managers) {
+          if (managers.size() == 0) {
+            throw std::runtime_error(
+                R"(There are no structure to get features from)");
+          }
+          auto n_neighbors{0};
+          for (auto & manager : managers) {
+            n_neighbors += manager->get_nb_clusters(2);
```

Comment:
how does this deal with half vs full neighbor list? The number of clusters of size 2 already deals with this distinction?

Reply:
by construction get_nb_clusters(2) will get the effective number of clusters. Then it's up to the user to know what is going on and adapt the code accordingly.

Reply:
I don't understand: is the effective number of cluster is the number of clusters as returned by the neighbor list, meaning `FullNL::get_nb_clusters(2) == 2 * HalfNL::get_nb_clusters(2)`? Or the actual number of different clusters, i.e. `FullNL::get_nb_clusters(2) == HalfNL::get_nb_clusters(2)`? 

As a Python user, I don't have control over the full or half neighbor list, so I can not really adapt the code here.

Reply:
`get_nb_clusters(2)` returns the number of clusters of order 2 in the current manager so for sure `FullNL::get_nb_clusters(2) != HalfNL::get_nb_clusters(2)`.
As a python user you know which type of manager you are working with, e.g. full or half NL, so when you work with the `dX_dr` with the `ij` mapping you know that things will probably need to be different depending on the type of NL you have been using. For instance in the tests I know that I have a full NL so I compute the kernel matrix elements with that in mind so the code would have to be modified a bit to properly handle a half NL.

I am not quite sure what is the problem here actually.

Reply:
The "standard" way of using rascal is the following, right?

```
frames = ase.io.read("file.xyz", ":")
soap = SphericalInvariants(...)

soap.transform(frames)
```

When I'm doing this, I have no idea is the code is using a full or half neighbors list, and I don't know how to control this parameter. I'd guess one kind of neighbor list is selected for me, but which one exactly is not documented nor guaranteed to stay the same.

The problem I see is that since I don't know which NL the code is using, I might write some code with one assumption, only to get wrong results later if my assumption of half/full NL turned out to be wrong.

Reply:
I see. You can't choose at the moment because half NL is not exposed to python. So this is something that we should keep in mind when this feature is made available but not at the moment. Also this kind of choice would be made explicit in the hypers of `SphericalInvariants`.

Reply:
These features are independent of the representation so to check which kind of NL is being used should look at the `AtomsList`. More specifically `AtomsList.nl_options` argument or the name of `AtomsList.managers` which are very explicit about this topic. So in your example doing:
```
frames = ase.io.read("file.xyz", ":")
soap = SphericalInvariants(...)
features = soap.transform(frames)
print(features.nl_options, features.managers) 
```
will give you well enough information. I am not sure how it could be easier to access this information?

Reply:
I did not knew this was possible, but it looks like it is not documented in https://cosmo-epfl.github.io/librascal/reference/python.html. 

Then I think this should work as is, provided we add the corresponding documentation to the website.